homeassistant/components/roon/event.py

```diff
@@ -0,0 +1,103 @@
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Roon Event from Config Entry."""
+    roon_server = hass.data[DOMAIN][config_entry.entry_id]
+    event_entities = set()
+
+    @callback
+    def async_add_roon_volume_entity(player_data):
+        """Add or update Roon event Entity."""
+        dev_id = player_data["dev_id"]
+        if dev_id in event_entities:
+            return
+        # new player!
+        event_entity = RoonEventEntity(roon_server, player_data)
+        event_entities.add(dev_id)
+        async_add_entities([event_entity])
+
+    # start listening for players to be added from the server component
+    async_dispatcher_connect(hass, "roon_media_player", async_add_roon_volume_entity)
+
+
+class RoonEventEntity(EventEntity):
+    """Representation of a Roon Event entity."""
+
+    _attr_device_class = EventDeviceClass.BUTTON
+    _attr_event_types = ["volume_up", "volume_down"]
+
+    def __init__(self, server, player_data):
+        """Initialize the entity."""
+        self._server = server
+        self._player_data = player_data
+        player_name = player_data["display_name"]
+        self._attr_name = f"{player_name} roon volume"
+
+    @property
+    def device_info(self) -> DeviceInfo | None:
+        """Return the device info."""
+        if self.unique_id is None:
+            return None
+        if self._player_data.get("source_controls"):
+            dev_model = self._player_data["source_controls"][0].get("display_name")
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.unique_id)},
+            # Instead of setting the device name to the entity name, roon
+            # should be updated to set has_entity_name = True, and set the entity
+            # name to None
+            name=cast(str | None, self.name),
+            manufacturer="RoonLabs",
+            model=dev_model,
+            via_device=(DOMAIN, self._server.roon_id),
+        )
+
+    @callback
+    def _roonapi_volume_callback(
+        self, control_key: str, event: str, value: int
+    ) -> None:
+        """Callbacks from the roon api with volume request."""
+
+        if event != "set_volume":
+            _LOGGER.debug("Received unsupported roon volume event %s", event)
+            return
+
+        if value > 0:
+            event = "volume_up"
+        else:
+            event = "volume_down"
+
+        self._trigger_event(event)
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Register volume hooks with the roon api."""
+
+        self._server.roonapi.register_volume_control(
```

Comment:
Do we need to unregister the callback when removing the entity?

Reply:
There isn't currently an pyroon api call to do this.

Pyroon is reverse engineered and when I investigated I couldn't find how to unregister these. I'm assuming it's possible although the official roon node api examples don't do so.

The pyroon mapping layer has a broad exception catch when it calls the callback - so if there is a problem it will catch it and report an error (although to date  I've never seen this)

From a roon perspective if the client goes away - then the volume hooks are removed at some point.

Reply:
I meant to remove the stored callback. Currently there's a memory leak here since the entity can be removed and added but the callback remains and whenever the entity is added, eg when changing the entity_id, a new callback will be stored. Generally there should be a way to remove callbacks if we can register callbacks.

Reply:
As I said there currently isn't a way to do this in the pyroon api.

All pyroon stores is the provided kay and the call back - so even if entities were being created and destroyed (which they generally aren't) it would be a very small growth in memory use

I guess it would be possible to remove the callback dictionary entry in oyroon - but I'm not convinced it's a great idea without also removing the matching entry via the roon_api..

There is a mechanism where roon to re-requests all the volume endpoints - which again makes me cautions about not having the two sides of the api match.

Reply:
Just double checked - and the `dev_id` is the same across runs - so I think this won't leak at all since it just reuses the same dictionary key.

Reply:
At the time of writing this is still using the entity_id as control_key.

Side note: Please don't resolve conversations if not addressing the comment according to the suggestion or if there's still disagreement.

Reply:
To be clear, whenever the user changes the entity_id, the entity will be removed and then added again.

Reply:
Apologies - neglected to commit and push the latest changes.

Reply:
I did some digging and guessing around the roon_api and managed to find out how to remove the volume hooks.

So have now released a new version of pyroon which has the ability to do this properly.

I've include the library upgrade (and included the release diff) here - but can break into another PR if you prefer.

I tested this by disabling the integration - and can confirm it works correctly removing the hooks in roon.